homeassistant/components/hunterdouglas_powerview/__init__.py

```diff
@@ -58,46 +41,63 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     config = entry.data
 
     hub_address = config[CONF_HOST]
+    api_version = config.get(CONF_API_VERSION, None)
+    _LOGGER.debug("Connecting %s at %s with v%s api", DOMAIN, hub_address, api_version)
+
     websession = async_get_clientsession(hass)
 
-    pv_request = AioRequest(hub_address, loop=hass.loop, websession=websession)
+    pv_request = AioRequest(
+        hub_address, loop=hass.loop, websession=websession, api_version=api_version
+    )
 
     try:
         async with asyncio.timeout(10):
-            device_info = await async_get_device_info(pv_request, hub_address)
+            hub = Hub(pv_request)
+            await hub.query_firmware()
+            device_info = await async_get_device_info(hub)
+            if hub.role != "Primary":
+                # this should be caught in config_flow, but account for a hub changing roles (which only happens manually by a user)
+                _LOGGER.error(
+                    "%s (%s) is performing role of %s Hub. Only the Primary Hub can manage shades",
+                    hub.name,
+                    hub.hub_address,
+                    hub.role,
+                )
+                return False
+            if CONF_API_VERSION not in config:
+                new_data = {**entry.data}
+                new_data[CONF_API_VERSION] = hub.api_version
+                hass.config_entries.async_update_entry(entry, data=new_data)
```

Comment:
Please move this block outside of the context manager since it is not expected to timeout